torch/csrc/jit/runtime/script_profile.cpp

```diff
@@ -0,0 +1,124 @@
+#include <torch/csrc/jit/runtime/script_profile.h>
+
+#include <atomic>
+#include <chrono>
+#include <mutex>
+#include <unordered_set>
+
+#include <c10/util/Exception.h>
+#include <torch/csrc/jit/api/function_impl.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+class ProfilesRegistry {
+ public:
+  bool empty() {
+    return empty_.load(std::memory_order_relaxed);
+  }
+  void addProfile(ScriptProfile& p) {
+    std::lock_guard<std::mutex> g(mutex_);
+    enabledProfiles_.emplace(&p);
+    empty_.store(false, std::memory_order_relaxed);
+  }
+  void removeProfile(ScriptProfile& p) {
+    std::lock_guard<std::mutex> g(mutex_);
+    enabledProfiles_.erase(&p);
+    if (enabledProfiles_.empty()) {
+      empty_.store(true, std::memory_order_relaxed);
+    }
+  }
+  void send(profiling::Datapoint datapoint) {
+    std::lock_guard<std::mutex> g(mutex_);
+    for (auto* p : enabledProfiles_) {
+      p->addDatapoint(datapoint);
+    }
+  }
+
+ private:
+  std::atomic<bool> empty_{true};
+  std::mutex mutex_;
+  std::unordered_set<ScriptProfile*> enabledProfiles_;
+};
+
+ProfilesRegistry& getProfilesRegistry() {
+  static auto& registry = *new ProfilesRegistry{};
+  return registry;
+}
+
+} // namespace
+
+namespace profiling {
+
+/* static */ c10::optional<InstructionSpan> InstructionSpan::tryMake(
+    Node& node) {
+  if (getProfilesRegistry().empty()) {
+    return {};
+  }
+  return InstructionSpan{node};
+}
+
+InstructionSpan::InstructionSpan(Node& node) : datapoint_(node.sourceRange()) {}
+
+InstructionSpan::InstructionSpan(InstructionSpan&& other)
+    : empty_(other.empty_), datapoint_(std::move(other.datapoint_)) {
+  other.empty_ = true;
+}
```

Comment:
Is this for moving `InstructionSpan`? Could we disallow moving to save the overhead of tracking/checking `empty_`? It would also reduce the cognitive load of knowing `c10::nullopt` is same as `InstructionSpan{empty_==False}`.

Reply:
yep, initially I thought the moving was unavoidable, but after second thoughts I think there's a legit trick in C++ to avoid the move. Will update.